synapse/lib/oauth.py

```diff
@@ -185,39 +158,108 @@ async def _oauthRefreshLoop(self):
         return await self._fetchOAuthToken(token_uri, auth, formdata, ssl_verify=ssl_verify)
 
-    async def _refreshOAuthAccessToken(self, providerconf, clientconf):
+    async def _refreshOAuthAccessToken(self, providerconf, clientconf, useriden):
+
+        ok, data = await self._getAuthData(providerconf, useriden)
+        if not ok:
+            return ok, data
+
         token_uri = providerconf['token_uri']
         ssl_verify = providerconf['ssl_verify']
         refresh_token = clientconf['refresh_token']
 
-        formdata = aiohttp.FormData()
+        auth = data.get('auth')
+        formdata = data.get('formdata')  # type: aiohttp.FormData
+
         formdata.add_field('grant_type', 'refresh_token')
         formdata.add_field('refresh_token', refresh_token)
 
-        auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
         ok, data = await self._fetchOAuthToken(token_uri, auth, formdata, ssl_verify=ssl_verify, retries=3)
         if ok and not data.get('refresh_token'):
             # if a refresh_token is not provided in the response persist the existing token
             data['refresh_token'] = refresh_token
 
         return ok, data
 
+    async def _getAuthData(self, providerconf, useriden):
+        isok = False
+        ret = {}
+        auth_scheme = providerconf['auth_scheme']
+
+        if auth_scheme == 'basic':
+            auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
+            ret['auth'] = auth
+            ret['formdata'] = aiohttp.FormData()
+            isok = True
+
+        elif auth_scheme == 'client_assertion':
+            assertion = None
+            client_id = providerconf['client_id']
+            client_assertion = providerconf['client_assertion']
+
+            if (info := client_assertion.get('cortex:callstorm')):
+                opts = {
+                    'view': info['view'],
+                    'vars': info.get('vars', {}),
+                    'user': useriden,
+                }
+                try:
+                    ok, info = await self.callStorm(info['query'], opts=opts)
+                except Exception as e:
+                    isok = False
+                    ret['error'] = f'Error executing callStorm: {e}'
+                else:
+                    if not ok:
+                        return ok, info
+                    assertion = info.get('token')
+
+            elif client_assertion.get('msft:azure:workloadidentity'):
+                ok, valu = _getAzureTokenFile()
+                if not ok:
+                    return ok, {'error': valu}
+                assertion = valu
+
+            else:  # pragma: no cover
+                isok = False
+                ret['error'] = f'unknown client_assertions data: {client_assertion=}'
+
+            if assertion:
+                formdata = aiohttp.FormData()
```

Comment:
Should this be a dict for potential serialization / portability or is having this be a heavy `aiohttp.FormData` advantageous?